### PR TITLE
Introduce fast PHP7 based linter

### PR DESCRIPTION
**Pull request by keradus**
Closes enhancement part of #1959 .

Impact only on PHP 7

Command: `php php-cs-fixer fix -v`

**master**:
> Fixed all files in 19.708 seconds, 8.000 MB memory used
> Fixed all files in 19.647 seconds, 8.000 MB memory used
> Fixed all files in 19.824 seconds, 8.000 MB memory used

**this PR**:
> Fixed all files in 13.238 seconds, 8.000 MB memory used
> Fixed all files in 13.227 seconds, 8.000 MB memory used
> Fixed all files in 13.227 seconds, 8.000 MB memory used

**~33 %** faster

### CODE REVIEW

**Comment by gharlan** [comment]
What is happening when disabling linter on php 7?
If I see it correctly the code is linted anyway, because of `TOKEN_PARSE`. But the exception for invalid files is not catched, because `TokenizerLinter` is not used.

**Comment by keradus** [comment]
Remember it's very early preview ;)
Indeed, on 7.x linter should be enforced.

**Comment by gharlan** [comment]
I still think linting should be always enforeced and we should never fix invalid files (especially files from higher php versions).

**Comment by keradus** [comment]
The problem is that for PHP5 you are not always able to enforce linting, as you need to find PHP executable path

**Comment by GrahamCampbell** [comment]
I often want to turn off linting so I can debug bad fixers.

**Comment by keradus** [comment]
one way or another for PHP7 linting would be always enabled thanks to Tokenizer change for semi-reserved keywords

**Comment by GrahamCampbell** [comment]
What's the status here?

**Comment by keradus** [comment]
I would guess it needs some dev time to be finished

File: src/Runner/Runner.php [hunk 1]
```diff
@@ -128,7 +128,7 @@ public function fix()
         $finder = $config->getFinder();
         $finderIterator = $finder instanceof \IteratorAggregate ? $finder->getIterator() : $finder;
 
-        $collection = new FileLintingIterator(
+        $collection = new FileCachingLintingIterator(
```
Comment (GrahamCampbell) [inline src/Runner/Runner.php]: Guess this is yet to be implemented?

File: src/Runner/FileCachingLintingIterator.php [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+use PhpCsFixer\Linter\LinterInterface;
+use PhpCsFixer\Linter\LintingResultInterface;
+
+/**
+ * @author Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ *
+ * @internal
+ */
+final class FileLintingIterator extends \CachingIterator
```
Comment (GrahamCampbell) [inline src/Runner/FileCachingLintingIterator.php]: Typo in the filename here. :)

**Comment by keradus** [comment]
cc @SpacePossum @GrahamCampbell

**Comment by keradus** [comment]
Also, updated initial PR comment

**Comment by GrahamCampbell** [comment]
Perhapse a linter factory would be better than having a linter wrapper class?

**Comment by keradus** [comment]
It was my first implementation. Then I got sick of it to replace the usage everywhere so I decided to kept the original name as is and encapsulate decision which linter will be use for component, so one could treat Process/Tokenizer Linters only as private classes for Linter namespace (as if it would be possible in PHP, like in C#)

**Pull request merged**